### PR TITLE
JP-3281: Change resampling defaults to NaN padding instead of INDEF

### LINKED ISSUE #7664: Is the fillval = 'INDEF' parameter in resample_spec working as intended?
_Issue [JP-3281](https://jira.stsci.edu/browse/JP-3281) was created on JIRA by [Kayli Glidic](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kglidic):_

Hi, 

I wanted to bring attention to a potential issue I encountered regarding one of the default parameters used in the resample_spec step of Stage 2. Upon analyzing NIRSpec Fixed Slit data, specifically S1600A1, which is observing background sky, I noticed a problem with the resulting data in the cal.fits file (from stage 2) vs. the resulting data in the s2d.fits file (from stage 2). 

 

Specifically, I observed that some pixels at the edges of the slit have values of NaN in the cal.fits file. However, when I examined the resulting s2d.fits file (also from stage 2), I found that those previously NaN pixels disappeared and seem to have been replaced with zeros. This replacement of NaN with zeros can be misleading once the pipeline gets to the extract_1d step, especially for extracting a background spectrum/extended source. Ideally, for extended sources, the extraction should encompass the entire slit, which means it encloses those zero-valued pixels at the edges in the s2d.fits data. Consequently, these zero-valued pixels are then factored into the final flux calculation, which makes me question the reliability of the resulting spectrum.

I only began to question my final spectrum after looking at the NPIXELS column in the 1D table. Looking at the pipeline documentation it states: "NPIXELS is the number of pixels that were added together for the source extraction region. Note that this is not necessarily a constant, and the value is not necessarily an integer (the data type is float)."

When I looked at my NPIXELS data, the value was always constant, it didn't vary by column which means it does not exclude those pixels with no flux. Columns with zero-valued pixels are then returning lower than expected surface brightness values. 

I did some digging and found that the parameter 'fillval' in resample_spec is set to 'INDEF' by default. Based on my understanding and best assumption, this fill value should assign NaN to pixels with no flux, which does not appear to be happening. As a quick test, I ran Stage 2 using both the default 'fillval' of 'INDEF' and by explicitly setting 'fillval' to NaN using the command spec2.resample_spec.fillval = 'nan' in my Jupyter notebook. I have attached the results to this ticket, including the resampled 2D spectra and the resulting 1D spectra for both cases. When I set 'fillval' to 'nan' explicitly, the pixels in the 2D spectra appear white with no assigned value, in contrast to the default 'fillval' where the pixels have a value of zero. The difference in 1D spectra between the two cases is also noticeable upon comparison which is why I wanted to bring this potential issue up. Also, just to note the same extraction region was used to get the 1D spectra in both cases. 

I am uncertain if this parameter is used in other steps, but at least in the context of the resample_spec step, this parameter causes a difference in my final results. I came across a GitHub issue from 2018 that addressed the use of the 'INDEF' fillval, but it still appears open and I didn't come across a specific JIRA ticket about this when I searched ([https://github.com/spacetelescope/jwst/issues/2219]). 

My concern is whether 'INDEF' is functioning as intended. Might this be a discrepancy between the expected behavior and the actual outcome when using this default?

### PR DESCRIPTION
**Pull request by drlaw1558**
Resolves [JP-3281](https://jira.stsci.edu/browse/JP-3281) by changing padding around science data in resampled arrays to NaN instead of INDEF.

Closes #7664 

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [x] All comments are resolved
- [x] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by drlaw1558** [comment]
Probably failing a bunch of reg tests as outputs are now NaN-padded insted of 0-padded.

File: CHANGES.rst [hunk 1]
```diff
@@ -130,11 +130,15 @@ resample
 - Update variance handling to propagate resampled variance components with
   weights that match the science `weight_type`. [#8437]
 
+- Change default no-value padding from INDEF to NaN [#8488]
```
Comment (hbushouse) [inline CHANGES.rst:133]:
```suggestion
- Change `fillval` parameter default from INDEF to NaN [#8488]
```

File: CHANGES.rst [hunk 2]
```diff
@@ -130,11 +130,15 @@ resample
...
+
 resample_spec
 -------------
 
 - Populate the wavelength array in resampled `Slit` and `MultiSlit` models. [#8374]
 
+- Change default no-value padding from INDEF to NaN [#8488]
```
Comment (hbushouse) [inline CHANGES.rst:140]:
```suggestion
- Change `fillval` parameter default from INDEF to NaN [#8488]
```

**Comment by hbushouse** [comment]
Regression tests started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1459

File: CHANGES.rst (hunk 1 above)
Comment (drlaw1558) [reply CHANGES.rst]: Done

File: CHANGES.rst (hunk 2 above)
Comment (drlaw1558) [reply CHANGES.rst]: Done

**Comment by drlaw1558** [comment]
> Unit tests contained in the jwst/resample/tests/test_resample_step.py module are causing all the CI failures, due to the change from zero to NaN in output arrays. These tests will need updating.

This seems a little tricky.  The failing test test_custom_refwcs_resample_imaging is creating an array of random numbers, passing it through resample in two slightly different ways, and checking that the results are allclose.  However, for whatever reason the values coming out of resample before this PR are all zeros everywhere in both resample calls.  All zeroes matches all zeroes.  After this change the results of both calls are two arrays of all NaNs.  All NaNs apparently is not considered a match for all NaNs.

The easiest thing to do is delete the assert np.allclose(data1, data2) as it doesn't seem to be testing anything meaningful anyway.  However, what was it meant to test?  Calling resample on the random number array ordinarily gives meaningful results, but the test as set up is trimming to an entirely empty part of the array.

**Comment by hbushouse** [comment]
> > Unit tests contained in the jwst/resample/tests/test_resample_step.py module are causing all the CI failures, due to the change from zero to NaN in output arrays. These tests will need updating.
> 
> This seems a little tricky. The failing test test_custom_refwcs_resample_imaging is creating an array of random numbers, passing it through resample in two slightly different ways, and checking that the results are allclose. However, for whatever reason the values coming out of resample before this PR are all zeros everywhere in both resample calls. All zeroes matches all zeroes. After this change the results of both calls are two arrays of all NaNs. All NaNs apparently is not considered a match for all NaNs.
> 
> The easiest thing to do is delete the assert np.allclose(data1, data2) as it doesn't seem to be testing anything meaningful anyway. However, what was it meant to test? Calling resample on the random number array ordinarily gives meaningful results, but the test as set up is trimming to an entirely empty part of the array.

You're right - that test is just bizarre. The nircam_rate image is defined to have a size of 204x204, and then the test is resampling to an output size of more than 1000x1000? And setting the wcs reference at around 600? Guessing that the 204x204 size might've been a mistake and that they really meant 2048x2048, I tried running it with that image size and the result (using the current master branch) is still zero-filled in all the SCI, WHT, and CON arrays, which means it has no input pixels contributing to any of the output pixels anywhere in the entire image. That whole test is just screwy. I suggest adding an xfail to it, in order to allow it to fail for now, and we'll try to get someone to fix/rewrite it in the future.

**Comment by hbushouse** [comment]
Started a regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1471

**Comment by drlaw1558** [comment]
> I suggest adding an xfail to it, in order to allow it to fail for now, and we'll try to get someone to fix/rewrite it in the future.

Done.

**Comment by drlaw1558** [comment]
> The latest regtest results seem pretty reasonable. Lots of i2d and s2d products that now have zero values changed to NaN, some of which then propagate into NaN values of x1d products (if they're spectra). In general it's 5-20% of the pixels that see such a change. There are a few cases, however, where as many as 80-90% of the pixels in an i2d or s2d have been changed from zero to NaN. This suggests we might have some fairly ratty test data in some tests, if that many pixels of the resample products have no contribution from their inputs. But mechanically it all looks reasonable.

@hbushouse Can you point me at a couple of examples with 80-90% NaN-valued i2d and s2d data?

**Pull request merged**